### ISSUE TITLE
moe_lora + use_vera adapts no expert on a fused-expert MoE: the flag loads, the MoE step runs, and VeRA trains attention only

### ISSUE THREAD

**Issue by Srinivasan8888** [MakazhanAlpamys/Soup#1197]
`training.moe_lora: true` combined with `training.lora.use_vera: true` loads, the trainer runs the MoE target step (it prints "ScatterMoE LoRA: targeting 7 module patterns"), and then **no expert is adapted**. The run trains attention-only VeRA under a config that asked for expert adapters. That is the "accepted, then silently not applied" class #1151 closes for tasks, here on one adapter type.

**Measured** on `main` `92bd2207`, CPU. I used the real `IPOTrainerWrapper._setup_transformers` (wired by #1148) on a ~10k-parameter Qwen3-MoE with fused 3-D experts, using the harness in `tests/test_issue1099_moe_lora_remaining_tasks.py`, which stubs only the loaders. I counted the PEFT tuner layers the attach created:

```
use_vera: false, moe_lora: true    LoRA layers 12, of them on mlp.experts: 4
use_vera: true,  moe_lora: false   VeRA layers  4, of them on mlp.experts: 0   (q/k/v/o)
use_vera: true,  moe_lora: true    VeRA layers  4, of them on mlp.experts: 0   <- flag had no effect
```

**Why.** For `use_vera`, `peft_wiring`'s adapter-kwargs helper returns a `VeraConfig` built from `target_modules` alone (the `use_vera` branch just before `_settle_unmapped`). On a fused-expert model, the experts are 3-D parameters (`mlp.experts.gate_up_proj`) that only the LoRA `target_parameters` / `ParamWrapper` path can address, which is why #798 routes `moe_lora` through it. VeRA has no equivalent, so the expert targets never reach it.

**Scope, stated so it is not over-read:**
- I measured one trainer and one fused architecture. Every MoE-wired trainer shares `build_lora_config`, so I expect the same on all of them, but I have not run each.
- A model with one `nn.Linear` per expert (the older Mixtral layout) may adapt fine under VeRA. I have not measured that, and it is the reason a config-load refusal would be too broad.
- Found by a local CodeRabbit review of #1179, and verified before filing. It is not introduced by #1179: the same result reproduces on `main`.

**Possible shapes** (your call):
- Refuse at the model, where fused vs per-module experts is known, as #798's dropout refusal does: `use_vera` + `moe_lora` on a model whose experts are fused parameters stops with a message naming both flags.
- Or refuse the combination at config load, if VeRA on MoE experts is not something Soup wants to support at all.

**Acceptance criteria**
- [ ] `use_vera: true` + `moe_lora: true` on a fused-expert model does not train attention-only in silence. A test that fails on today's `main` counts the tuner layers on `mlp.experts`.
- [ ] Whatever path is chosen, a dense model and a `use_vera`-only config are unchanged (controls).

I am not claiming this yet. I have four PRs in review and want to finish those first. If nobody takes it by then, I will ask.

**Comment by MakazhanAlpamys** [comment]
@Srinivasan8888, thanks for measuring this before filing. It reproduces.

**Measured on `main` `210298c8`** (transformers 5.17.0, peft 0.20.0, CPU). I ran the real `IPOTrainerWrapper._setup_transformers` and `DPOTrainerWrapper._setup_transformers` with only the loaders stubbed, on the same ~10k-parameter fused-expert Qwen3-MoE, and counted peft tuner layers (`BaseTunerLayer` modules):

```
                             ipo                      dpo
LoRA, moe_lora: true         6 layers, 2 on experts   6 layers, 2 on experts
VeRA, moe_lora: false        4 layers, 0 on experts   4 layers, 0 on experts
VeRA, moe_lora: true         4 layers, 0 on experts   4 layers, 0 on experts   <- flag has no effect
```

Your 12 / 4 counts `lora_A` names, which appear twice per adapted layer (`...lora_A` and `...lora_A.default`), so it is the same result. The VeRA + `moe_lora` runs print "ScatterMoE LoRA: targeting 7 module patterns" and then adapt the same four attention projections as VeRA without the flag.

`dpo` has read the flag since #798, so this predates both #1148 and #1179.

**The cause is as you describe.** `build_peft_config_spec` (`utils/peft_wiring.py:438-448`) builds `VeraConfig` from `target_modules` alone. VeRA wraps `nn.Linear` layers only. Expert patterns that name fused 3-D parameters are skipped without a word, and peft raises only when nothing matches at all.

**One data point that bears on the fix shape.** On transformers 5.17.0 the in-library Mixtral is fused as well: `MixtralExperts` holds `gate_up_proj` / `down_proj` as parameters. There, LoRA + `moe_lora` and VeRA + `moe_lora` both give 4 tuner layers and 0 on experts. The LoRA half is the documented "`moe_lora` does not reach every MoE family" limitation. So the one-`nn.Linear`-per-expert layout you mention does not occur for Mixtral at Soup's transformers floor. Also, #798's predicate `peft_routes_lora_to_fused_params` is false for Mixtral, so a VeRA check keyed to that predicate would not fire there. Whoever takes this should choose between two checks:

- "VeRA wraps no expert module", which also refuses Mixtral;
- reusing #798's predicate, which leaves Mixtral where LoRA already is: it attaches and adapts no experts.

**This stays separate from #1179.** The natural home for the check is `resolve_moe_lora_targets` (`utils/moe.py:197`), next to `_refuse_dropout_on_fused_experts` (`:176`), because that is where the loaded model is known. Every MoE-wired trainer calls that helper, so one check covers all of them, including the three that #1179 wires. #1179 does not touch that file, so there is no ordering constraint between the two. Your acceptance criteria are the right ones: a test that fails on today's `main` by counting tuner layers on `mlp.experts`, plus the dense and VeRA-only controls.

**My call on the shape: the first check.** Refuse when VeRA wraps no expert module, because then `moe_lora` does nothing on any model, and Soup refuses a setting it cannot honour rather than ignoring it. That it also refuses VeRA on Mixtral is correct for the same reason; LoRA on Mixtral stays the separately documented limitation.

It is unclaimed: no assignee, and you said you are not taking it yet. Anyone who wants it, please comment here first.

**Comment by Srinivasan8888** [comment]
Thank you for reproducing it and for the correction. You're right about my counts: 12 / 4 counted `lora_A` names, which appear twice per adapted layer, so the true figures are 6 layers, 2 on experts, as in your table. My note that per-expert Mixtral "may adapt fine" was speculation, and your measurement shows Mixtral is fused at the transformers floor too. The fix shape you ruled (refuse when VeRA wraps no expert module) covers both. I'm still not claiming it until my open PRs land.
